### ISSUE TITLE
Consider renaming top-level select to select_

### ISSUE THREAD

**Issue by trevorbaca** [Abjad/abjad#635]
Turns out that `select` is, in fact, the name of a module in Python's standard library:

```
Python 2.7.10 (default, Oct 23 2015, 19:19:21) 
[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.59.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import select
>>> dir(select)
['KQ_EV_ADD', 'KQ_EV_CLEAR', 'KQ_EV_DELETE', 'KQ_EV_DISABLE', 'KQ_EV_ENABLE', 'KQ_EV_EOF', 'KQ_EV_ERROR', 'KQ_EV_FLAG1', 'KQ_EV_ONESHOT', 'KQ_EV_SYSFLAGS', 'KQ_FILTER_AIO', 'KQ_FILTER_PROC', 'KQ_FILTER_READ', 'KQ_FILTER_SIGNAL', 'KQ_FILTER_TIMER', 'KQ_FILTER_VNODE', 'KQ_FILTER_WRITE', 'KQ_NOTE_ATTRIB', 'KQ_NOTE_CHILD', 'KQ_NOTE_DELETE', 'KQ_NOTE_EXEC', 'KQ_NOTE_EXIT', 'KQ_NOTE_EXTEND', 'KQ_NOTE_FORK', 'KQ_NOTE_LINK', 'KQ_NOTE_LOWAT', 'KQ_NOTE_PCTRLMASK', 'KQ_NOTE_PDATAMASK', 'KQ_NOTE_RENAME', 'KQ_NOTE_REVOKE', 'KQ_NOTE_TRACK', 'KQ_NOTE_TRACKERR', 'KQ_NOTE_WRITE', 'PIPE_BUF', '__doc__', '__file__', '__name__', '__package__', 'error', 'kevent', 'kqueue', 'select']
```

As a rule we've tried to avoid Abjad aliasing anything in the standard library. (Which is why, for example, our top-level `inspect_` function is named the way that it is.)

The reason I raise the point now is that I recently added a `select.py` module to one of my own projects and then later ran into exactly the same type of aliasing drama we experienced earlier when our `inspect_.py` module was named `inspect.py`. So I renamed my `select.py` module to `select_.py` and everything was happy.

However, it's also true that we haven't (yet) encountered select-aliasing in Abjad (or at least it hasn't been reported if our users have experienced it; but I think it's likely that select-usage hasn't made it very far into the userbase yet). So perhaps it's a problem that won't ever appear. Though I suspect that, at some point, we'll get bit in a deep way by aliasing when we least expect it.

Cons: user-visible change will have to be communicated clearly.

Pros: naming consistency; heads off longterm problem that may force us to make a user-facing change when we're unprepared.

Other thoughts?

**Comment by josiah-wolf-oberholtzer** [comment]
I think the chance of an Abjad user using the built-in `select` library is nearly 0%.

There's a good reason (even if annoying) reason that we renamed `inspect` to `inspect_`: the built-in `inspect` module is super useful and often germane to work you do in complex, dynamic score-generation.

But the `select` module is for low-level systems programming. We don't even use it anywhere in our codebase (like we do with the built-in `inspect`) because there are plenty of nice higher-level tools for dealing with processes and pipes such as `subprocess.call()`.

I vote :-1:.

As an aside, this whole issue of (renaming to avoid) aliasing built-ins would be moot if we accepted the community standard of not using `import *` everywhere.

**Comment by trevorbaca** [comment]
!

Where do we use `import *` everywhere?

**Comment by josiah-wolf-oberholtzer** [comment]
I mean `from abjad import *`.

**Comment by josiah-wolf-oberholtzer** [comment]
We're going to keep having this discussion about aliasing as long as we keep using `from abjad import *` everywhere. It's a conversation about namespacing, and `abjad` is not a very long namespace to type out. There will always be new aliases to discover. Do we keep breaking backwards compatibility because we find a name that clashes? Or do we acknowledge we have our own sovereign namespace to work within that doesn't require us to avoid clashes in the first place?

**Comment by trevorbaca** [comment]
What am I missing? Because it looks like all starred imports are exclusively in demos and test files. In fact none in our source files:
```
15:19 $ ajk 'from abjad import \*' | grep -v test

boilerplate/example_score/example_score/__metadata__.py:2:from abjad import *
boilerplate/example_score/example_score/materials/__abbreviations__.py:2:from abjad import *
boilerplate/example_score/example_score/materials/__views__.py:3:from abjad import *
boilerplate/example_score/example_score/segments/__views__.py:3:from abjad import *
demos/desordre/make_desordre_cell.py:3:from abjad import *
demos/desordre/make_desordre_measure.py:2:from abjad import *
demos/desordre/make_desordre_score.py:2:from abjad import *
demos/desordre/make_desordre_staff.py:2:from abjad import *
demos/ferneyhough/configure_lilypond_file.py:2:from abjad import *
demos/ferneyhough/configure_score.py:2:from abjad import *
demos/ferneyhough/make_lilypond_file.py:2:from abjad import *
demos/ferneyhough/make_nested_tuplet.py:2:from abjad import *
demos/ferneyhough/make_score.py:2:from abjad import *
demos/mozart/make_mozart_lilypond_file.py:2:from abjad import *
demos/mozart/make_mozart_measure.py:2:from abjad import *
demos/mozart/make_mozart_score.py:2:from abjad import *
demos/part/add_string_music_to_score.py:3:from abjad import *
demos/part/apply_bowing_marks.py:3:from abjad import *
demos/part/apply_dynamics.py:2:from abjad import *
demos/part/apply_expressive_marks.py:2:from abjad import *
demos/part/apply_final_bar_lines.py:2:from abjad import *
demos/part/apply_page_breaks.py:2:from abjad import *
demos/part/apply_rehearsal_marks.py:2:from abjad import *
demos/part/configure_lilypond_file.py:2:from abjad import *
demos/part/configure_score.py:2:from abjad import *
demos/part/create_pitch_contour_reservoir.py:2:from abjad import *
demos/part/durate_pitch_contour_reservoir.py:2:from abjad import *
demos/part/edit_cello_voice.py:2:from abjad import *
demos/part/edit_first_violin_voice.py:2:from abjad import *
demos/part/edit_second_violin_voice.py:2:from abjad import *
demos/part/edit_viola_voice.py:2:from abjad import *
demos/part/make_part_lilypond_file.py:2:from abjad import *
demos/part/PartCantusScoreTemplate.py:2:from abjad import *
scr/devel/prime-parser-tables:3:from abjad import *
tools/lilypondparsertools/LilyPondGrammarGenerator.py:168:            f.write('from abjad import *\n')
tools/rhythmmakertools/gallery/gallery.py:4:from abjad import *
tools/stringtools/format_input_lines_as_doc_string.py:40:    header = 'from abjad import *\n'
tools/stringtools/format_input_lines_as_doc_string.py:51:    header = 'from abjad import *\n'
tools/systemtools/run_abjad.py:19:        "from abjad import *;",
abjad-ack is running this command:

    ack --ignore-dir=docs --ignore-dir=build --ignore-file=is:.coverage 'from abjad import \*'


abjad-ack total time: 1 second(s).
```
I don't remember us ever allowing starred-import in the actual source files.

**Comment by josiah-wolf-oberholtzer** [comment]
No, in all doctests, all examples in the documentation, and the expectation setup in our docs that users should use `from abjad import *` when creating their own composition modules.

If we didn't encourage the `from abjad import *` construct, aliasing would be a non-issue. Aliases are only a problem when you actually overwrite one name with another. But if all of our names are kept in the `abjad` namespace in all examples, there's no reason to worry about `select`. There's no reason to worry about `inspect` either.

**Comment by trevorbaca** [comment]
I see. Yup, I'm in favor of recommending folks move away from starred imports in their own composition environments, too.

But what to do when someone *does* call `from abjad import *`? We can't physically tie users' hands (at least not all the time, anyway ;) And it feels like the system probably shouldn't outright break.

I think the principle should be something like "discourage starred imports, but make sure nothing breaks on the off occasion where ppl use one." But leaving `select` as it is fits this principle just fine; I raised the issue to give us the chance to have the discussion, because we have one function (`inspect_`) named according to one convention and another function (`select`) named according to another.

As an aside, I know that the Python community as a whole recommends avoidance of starred imports; this isn't news to me. The problem arises because Python continues to support the feature while simultaneously discouraging its use: when we see `__all__` removed from the language, we'll know that the community has truly migrated away from use in a practical sense, not just by convention.

**Comment by josiah-wolf-oberholtzer** [comment]
OK, good, agreed.

We can't prevent users from clobbering names. It's not possible. But I think we can pick our battles. 

Clobbering built-ins is a big no-no, and that's why I'll never complain about our `topleveltools.set_()`. But our users are far less likely to use the stdlib `inspect`, and will probably never ever use the stdlib `select`.

Note too in all of the discussions surrounding the use of trailing underscores in Python that people are always talking about aliasing built-ins and keywords, and rarely if ever about stdlib modules. I think clobbering stdlib modules is actually a fairly low-impact problem.

**Comment by trevorbaca** [comment]
Yes, yes, good, good.

The only problem is that stdlib module naming was enough of a problem to make us change `inspect_`. So it's evidently somewhat higher-impact than we'd like it to be! ;)

But I'm fine with leaving `select` as-is. So long as we acknowledge that we've now got two competing naming conventions as a result. (Which is basically something like "we add a trailing underscore for built-ins and also for some, but not all, standard library modules.) Of course that's a totally reasonable way to go: there are considerably worse naming conventions out there ;)

I'd actually be interested to know if he could retrovert `inspect_` back to just unmodified `inspect`. Then the naming convention is completely clear: underscores for builtins only.

**Comment by josiah-wolf-oberholtzer** [comment]
I'll go there.

I think the name real estate occupied by `inspect` is so good, so valuable, and our own inspection logic more useful and relevant  to our domain by two orders of magnitude over the stdlib `inspect` that we should seize the name via eminent domain.

Ask yourself, how often have you needed to use both `inspect` and `inspect_` in the same lexical scope?

**Comment by trevorbaca** [comment]
Absolutely! Couldn't agree more strongly. I hated ceding `inspect`! ;)

Ok, IIRC the reason we added the trailing underscore was because vanilla `inspect` caused Abjad not to start. So we'll have to make the change and test to see if that's still the case. (In other words, I think the problem arose completely independently of any user's practice in their own composition environment(s).)

I'll put it on a to-do list for the distant future ;)
